### PR TITLE
Added instance trait and simplified save/load logic for Dataset

### PR DESCRIPTION
**Pull request by nishaq503**
Also:

- Added Index trait on Dataset. This allows for square-bracket indexing.
- Cleaning up usage of cluster indices
- Renamed Dataset reordering to Dataset permutation.
- Removed one generic bound from Cluster
- Removed the `codec` module. This will come back when we start the implementation of compression.
- Updated the logic for converting between the name and history of a cluster.
- Updated two README files to not have to keep them in sync.
- Changed the default KNN algorithm to `GreedySieve`.

I will bump the version after the two reviews are done.

### CODE REVIEW

**Review by olwmc — APPROVED** [review]
(I only read the changes to code I personally worked on)

Excellent. Very good abstractions. Ship it.

**Review by olwmc — APPROVED** [review]
Good work. Sorry this took so long for me to review. No real comments. Very nice we can get rid of sized on instance.

File: abd-clam/src/core/dataset/vec2d.rs [hunk 1]
```diff
@@ -43,284 +43,236 @@ impl<T: Send + Sync + Copy, U: Number> VecDataset<T, U> {
...
+        self.metric
+    }
 
-        // Write null terminator
-        handle.write(&[0]).map_err(|err| format!("{err}"))?;
+    fn permute_instances(&mut self, permutation: &[usize]) -> Result<(), String> {
+        let n = permutation.len();
+
+        // The "source index" represents the index that we hope to swap to
+        let mut source_index: usize;
```
Comment (morganprior) [inline abd-clam/src/core/dataset/vec2d.rs:93]: This naming is counterintuitive to me

Comment (morganprior) [reply abd-clam/src/core/dataset/vec2d.rs:93]: Eh, I guess it's fine. It's a "source" index in the sense that it's the source of the element that will be put in position i on this iteration

File: abd-clam/src/core/dataset/vec2d.rs [hunk 2]
```diff
@@ -43,284 +43,236 @@ impl<T: Send + Sync + Copy, U: Number> VecDataset<T, U> {
...
-            .write_all(V::type_name().as_bytes())
-            .map_err(|err| format!("{err}"))?;
+    fn metric(&self) -> fn(&I, &I) -> U {
+        self.metric
+    }
 
-        // Write null terminator
-        handle.write(&[0]).map_err(|err| format!("{err}"))?;
+    fn permute_instances(&mut self, permutation: &[usize]) -> Result<(), String> {
+        let n = permutation.len();
```
Comment (morganprior) [inline abd-clam/src/core/dataset/vec2d.rs:90]: This came out really slick

File: abd-clam/src/core/cluster/_cluster.rs [hunk 3]
```diff
@@ -444,100 +445,57 @@ impl<T: Send + Sync + Copy, U: Number> Cluster<T, U> {
...
-                unreachable!("We know the characters used are only \"0\" and \"1\".")
-            };
-            acc.push_str(&format!("{s:01x}"));
-            acc
-        })
-    }
-
-    /// Converts the hexidecimal representation of cluster history obtained from `Cluster::name`
-    /// back into a `Vec<bool>`
+        let rem = history.len() % 4;
```
Comment (morganprior) [inline abd-clam/src/core/cluster/_cluster.rs:451]: This is really nice

**Review by morganprior — APPROVED** [review]
Looks good to me. This really neatened up function signatures. 

6341c8952e381bf138a39bb2cb0df03f

**Pull request merged**